libuavcan/include/libuavcan/platform/memory.hpp

```diff
@@ -304,6 +304,100 @@ class LIBUAVCAN_EXPORT PoolAllocator
     }
 };
 
+/**
+ * Copy bits from a byte array using arbitrary alignment to an aligned byte array.
+ *
+ * @param  src              The byte array to copy from.
+ * @param  src_offset_bits  The offset, in bits, from the start of the src array to
+ *                          start copying from.
+ * @param  dst              The byte array to copy data into.
+ * @param  length_bits      The total length of bits to copy. The caller must ensure
+ *                          that the size of src and dst are >= this value.
+ *
+ * @return The number of bits copied.
+ */
+inline std::size_t copyBitsUnalignedToAligned(const std::uint8_t* const src,
+                                              const std::size_t         src_offset_bits,
+                                              std::uint8_t* const       dst,
+                                              const std::size_t         length_bits)
+{
+    if (nullptr == src || nullptr == dst || length_bits == 0)
+    {
+        return 0;
+    }
+    std::size_t       bits_copied  = 0;
+    std::size_t       offset_bits  = src_offset_bits;
+    const std::size_t local_offset = src_offset_bits % 8U;
+    do
+    {
+        std::size_t       current_byte       = offset_bits / 8U;
+        const std::size_t bits_from_src_byte = 8U - local_offset;
+        bits_copied += std::min(length_bits, bits_from_src_byte);
+        dst[current_byte] &= static_cast<std::uint8_t>(0xFF << bits_from_src_byte);
+        dst[current_byte] |= static_cast<std::uint8_t>(src[current_byte] >> local_offset);
+        offset_bits += 8U;
+        if (offset_bits < length_bits)
+        {
+            current_byte      = offset_bits / 8U;
+            dst[current_byte] = static_cast<std::uint8_t>(src[current_byte] << bits_from_src_byte);
+            bits_copied += local_offset;
+        }
+        else
+        {
+            // we don't need to reevaluate the while condition.
+            break;
+        }
+    } while (true);
```

Comment:
Any reason to avoid `for (;;)`?

Reply:
You propose we lie with the emptiness ; between ; the semicolons ; ; the void that questions all symbols ; that screams with the horror of its own, undefined nature ;::.;.?!|

Nay, sir! I embrace the life-affirming : (all things are true)

Cynic.

Just a preference. It's the exact same set of instructions ether way.

Reply:
Convincing. Is there also a pearl of unspoken wisdom about the preference for `do ... while (true)` instead of `while (true)` that I should know about?

Reply:
nope. My weird psychology around this is just that, since we want the loop evaluation to go away, it's best to hide it at the bottom, so only human factors are involved. I'm also one of those people that put private stuff in classes at the bottom of the definition but I've had to change my ways for Zubax style!

Reply:
People following the Zubax style should be thankful for saving one `private` per class definition.

![image](https://user-images.githubusercontent.com/3298404/62794082-decb9980-badb-11e9-9052-4764985f27d5.png)